File: skills/ecological-data-foundation/scripts/download_from_obis.py
"""Download marine occurrence records from OBIS (Ocean Biodiversity Information System).

Usage: python download_from_obis.py <species_name_or_list_csv> <output_dir> [year_from] [year_to] [wkt_geometry]

Arguments:
    species_name_or_list_csv : Species name (e.g., "Chelonia mydas") or path to CSV
                               with column 'scientificName'
    output_dir               : Directory for outputs (created if absent)
    year_from                : Minimum year of observation (default: 1950)
    year_to                  : Maximum year of observation (default: current year)
    wkt_geometry             : WKT polygon to restrict query (optional)

Outputs (per species):
    occurrences_raw_OBIS_{species}_{date}.csv  — standardised occurrence records
    download_metadata_OBIS_{species}.txt        — download provenance and citation

Standard output schema:
    species, decimalLatitude, decimalLongitude, eventDate, countryCode,
    basisOfRecord, coordinateUncertaintyInMeters, datasetName, occurrenceID,
    source, download_doi
Extra OBIS columns:
    depth, marine
"""

import logging
import sys
from datetime import datetime
from pathlib import Path

SKILL_NAME = "ecological-data-foundation"
_LOG_DIR   = Path("logs")
_LOG_DIR.mkdir(parents=True, exist_ok=True)
_log_file  = _LOG_DIR / f"skill_{SKILL_NAME}_{datetime.now().strftime('%Y%m%d_%H%M%S')}.log"
logging.basicConfig(
    level=logging.INFO,
    format="[%(asctime)s] [%(levelname)s] [" + SKILL_NAME + "] %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
    handlers=[
        logging.StreamHandler(sys.stdout),
        logging.FileHandler(_log_file, encoding="utf-8"),
    ],
)
logger = logging.getLogger(SKILL_NAME)

def log_step(n: int, desc: str) -> None:
    logger.info("-- STEP %d: %s", n, desc)

def log_decision(var: str, val, why: str) -> None:
    logger.info("DECISION | %s = %s | %s", var, val, why)


from datetime import date

try:
    import requests
    import pandas as pd
except ImportError as e:
    logger.error(
        "Dependencia ausente: %s\n  Instale com: pip install requests pandas\n  Skill anterior: ecological-data-foundation",
        e,
    )
    sys.exit(1)


# ── Constants ─────────────────────────────────────────────────────────────────
OBIS_API_BASE = "https://api.obis.org/v3"
PAGE_SIZE     = 5000  # OBIS max per request
BAD_FLAGS     = {"NO_COORD", "ZERO_COORD", "ON_LAND", "DEPTH_EXCEEDS_BATH"}


# ── Helper functions ──────────────────────────────────────────────────────────

def fetch_obis(scientificname: str, year_from: int, year_to: int,
               wkt_geometry: str | None) -> list[dict]:
    """Fetch all OBIS records for a species, handling pagination."""
    url    = f"{OBIS_API_BASE}/occurrence"
    offset = 0
    all_results: list[dict] = []

    while True:
        params: dict = {
            "scientificname": scientificname,
            "absence":        "exclude",
            "startdate":      f"{year_from}-01-01",
            "enddate":        f"{year_to}-12-31",
            "size":           PAGE_SIZE,
            "offset":         offset,
        }
        if wkt_geometry:
            params["geometry"] = wkt_geometry

        try:
            resp = requests.get(url, params=params, timeout=60)
            resp.raise_for_status()
            data = resp.json()
        except requests.RequestException as e:
            logger.error(
                "Falha na requisicao OBIS (offset=%d, especie='%s'): %s\n  Causa provavel: sem conexao com a internet ou API OBIS indisponivel.\n  Verifique: https://api.obis.org/\n  Skill anterior: ecological-data-foundation",
                offset, scientificname, e,
            )
            raise

        results = data.get("results", [])
        if not results:
            break

        all_results.extend(results)
        logger.info("Pagina offset=%d: %d registros recuperados (total: %d)",
                    offset, len(results), len(all_results))

        total = data.get("total", 0)
        if len(all_results) >= total:
            break
        if len(results) < PAGE_SIZE:
            break
        offset += PAGE_SIZE

    return all_results


def apply_quality_flags(records: list[dict]) -> list[dict]:
    """Remove records with known OBIS quality issues."""
    filtered = []
    n_flagged = 0
    for rec in records:
        flags_raw = rec.get("flags", "") or ""
        flags = set(str(flags_raw).upper().split(","))
        if flags & BAD_FLAGS:
            n_flagged += 1
            continue
        filtered.append(rec)
    if n_flagged > 0:
        logger.warning("%d registros removidos por flags OBIS de qualidade.", n_flagged)
    return filtered


def standardise_records(records: list[dict], species_name: str) -> "pd.DataFrame":
    """Convert OBIS record dicts to the standard occurrence schema."""
    rows = []
    for rec in records:
        rows.append({
            "species":                       species_name,
            "decimalLatitude":               rec.get("decimalLatitude"),
            "decimalLongitude":              rec.get("decimalLongitude"),
            "eventDate":                     rec.get("eventDate") or rec.get("date_start", ""),
            "countryCode":                   rec.get("countryCode", ""),
            "basisOfRecord":                 rec.get("basisOfRecord", "OCCURRENCE"),
            "coordinateUncertaintyInMeters": rec.get("coordinateUncertaintyInMeters"),
            "datasetName":                   rec.get("datasetName", ""),
            "occurrenceID":                  str(rec.get("occurrenceID", rec.get("id", ""))),
            "source":                        "OBIS",
            "download_doi":                  None,
            "depth":                         rec.get("depth"),
            "marine":                        True,
        })

    df = pd.DataFrame(rows)
    df["decimalLatitude"]  = pd.to_numeric(df["decimalLatitude"],  errors="coerce")
    df["decimalLongitude"] = pd.to_numeric(df["decimalLongitude"], errors="coerce")
    df = df.dropna(subset=["decimalLatitude", "decimalLongitude"])
    return df.reset_index(drop=True)


def save_metadata(output_dir: Path, species_name: str, n_records: int,
                  year_from: int, year_to: int, wkt_geometry: str | None) -> None:
    safe_name = species_name.replace(" ", "_")
    today = date.today().isoformat()
    year  = date.today().year
    lines = [
        f"Species: {species_name}",
        f"Source: Ocean Biodiversity Information System (OBIS) — https://obis.org",
        f"API endpoint: {OBIS_API_BASE}/occurrence",
        f"Absence records excluded: True",
        f"OBIS quality flags applied: True (NO_COORD, ZERO_COORD, ON_LAND, DEPTH_EXCEEDS_BATH removed)",
        f"Year range: {year_from} - {year_to}",
        f"WKT geometry: {wkt_geometry or 'none (global)'}",
        f"n_records: {n_records}",
        f"Download date: {today}",
        (f"Citation: OBIS ({year}) Ocean Biodiversity Information System. "
         "Intergovernmental Oceanographic Commission of UNESCO. "
         f"www.obis.org. Accessed on {today}."),
        "License: CC0 1.0 (https://creativecommons.org/publicdomain/zero/1.0/)",
    ]
    meta_path = output_dir / f"download_metadata_OBIS_{safe_name}.txt"
    try:
        meta_path.write_text("\n".join(lines), encoding="utf-8")
        logger.info("Metadados gravados: %s", meta_path)
    except OSError as e:
        logger.error(
            "Falha ao gravar metadados: %s\n  Skill anterior: ecological-data-foundation", e,
        )
        raise


# ── Main download logic ────────────────────────────────────────────────────────

def download_species(species_name: str, output_dir: Path,
                     year_from: int, year_to: int,
                     wkt_geometry: str | None) -> None:
    logger.info("--- Iniciando download OBIS: %s ---", species_name)
    today_str = date.today().strftime("%Y%m%d")
    safe_name = species_name.replace(" ", "_")

    log_step(1, f"Buscar registros OBIS para '{species_name}'")
    records = fetch_obis(species_name, year_from, year_to, wkt_geometry)
    logger.info("Registros brutos recuperados: %d", len(records))

    if not records:
        logger.warning("Nenhum registro OBIS encontrado para '%s'.", species_name)
        return

    log_step(2, "Aplicar filtros de qualidade OBIS")
    records = apply_quality_flags(records)
    logger.info("Registros apos filtros OBIS: %d", len(records))

    log_step(3, "Padronizar registros para schema de saida")
    df = standardise_records(records, species_name)
    n_final = len(df)
    logger.info("Registros com coordenadas validas: %d", n_final)

    if n_final < 30:
        logger.warning(
            "Registros insuficientes para analise confiavel (n = %d). Considere relaxar filtros.",
            n_final,
        )

    log_step(4, "Gravar CSV de ocorrencias")
    csv_path = output_dir / f"occurrences_raw_OBIS_{safe_name}_{today_str}.csv"
    try:
        df.to_csv(csv_path, index=False)
        logger.info("Gravado: %s (%d registros)", csv_path, n_final)
    except OSError as e:
        logger.error(
            "Falha ao gravar CSV: %s\n  Skill anterior: ecological-data-foundation", e,
        )
        raise

    log_step(5, "Gravar metadados do download")
    save_metadata(output_dir, species_name, n_final, year_from, year_to, wkt_geometry)


# ── Entry point ────────────────────────────────────────────────────────────────

def main():
    logger.info("Script: download_from_obis.py | Skill: %s", SKILL_NAME)

    argv = sys.argv[1:]

    if len(argv) < 2:
        species_input = "Chelonia mydas"
        output_dir    = Path("output/obis")
        year_from     = 1950
        year_to       = date.today().year
        wkt_geometry  = None
        logger.warning("Menos de 2 argumentos fornecidos. Usando valores padrao para teste.")
    else:
        species_input = argv[0]
        output_dir    = Path(argv[1])
        year_from     = int(argv[2]) if len(argv) >= 3 else 1950
        year_to       = int(argv[3]) if len(argv) >= 4 else date.today().year
        wkt_geometry  = argv[4] if len(argv) >= 5 and argv[4] else None

    logger.info("Species input  : %s", species_input)
    logger.info("Output dir     : %s", output_dir)
    logger.info("Year range     : %d - %d", year_from, year_to)
    logger.info("WKT geometry   : %s", wkt_geometry or "nenhum (global)")

    log_decision("absence", "exclude",
                 "apenas registros de presenca confirmada")
    log_decision("quality_flags", str(BAD_FLAGS),
                 "registros com flags de qualidade OBIS sao removidos")

    output_dir.mkdir(parents=True, exist_ok=True)

    # Build species list
    log_step(0, "Construir lista de especies")
    if species_input.endswith(".csv") and Path(species_input).exists():
        try:
            df_sp = pd.read_csv(species_input)
            if "scientificName" not in df_sp.columns:
                logger.error(
                    "Coluna 'scientificName' nao encontrada em: %s\n  Skill anterior: ecological-data-foundation",
                    species_input,
                )
                sys.exit(1)
            species_list = df_sp["scientificName"].dropna().unique().tolist()
            logger.info("Modo batch: %d especies carregadas", len(species_list))
        except Exception as e:
            logger.error(
                "Falha ao ler lista de especies: %s\n  Skill anterior: ecological-data-foundation", e,
            )
            sys.exit(1)
    else:
        species_list = [species_input.strip()]
        logger.info("Modo especie unica: %s", species_list[0])

    for sp in species_list:
        try:
            download_species(sp, output_dir, year_from, year_to, wkt_geometry)
        except FileNotFoundError as e:
            logger.error(
                "Arquivo de entrada nao encontrado ao processar '%s': %s\n  Skill anterior: ecological-data-foundation",
                sp, e,
            )
        except Exception as e:
            logger.error(
                "Falha ao baixar '%s' do OBIS: %s\n  Causa provavel: problema de rede ou especie nao encontrada.\n  Skill anterior: ecological-data-foundation",
                sp, e,
            )

    logger.info("Todos os downloads OBIS concluidos. Verifique: %s", output_dir)


if __name__ == "__main__":
    main()

File: skills/ecological-data-foundation/scripts/download_from_obis.R
# ecological-agent-skills / Copyright (C) 2026 Francisco Diego Barros Barata
# SPDX-License-Identifier: GPL-3.0-or-later

# Usage: Rscript download_from_obis.R <species_name_or_list_csv> <output_dir> [year_from] [year_to] [wkt_geometry]

# ── Inline logger ─────────────────────────────────────────────────────────────
SKILL_NAME <- "ecological-data-foundation"
.log_ts  <- function() format(Sys.time(), "[%Y-%m-%d %H:%M:%S]")
log_info <- function(...) message(.log_ts(), " [INFO]  ", sprintf(...))
log_warn <- function(...) message(.log_ts(), " [WARN]  ", sprintf(...))
log_error<- function(...) message(.log_ts(), " [ERROR] ", sprintf(...))
log_step <- function(n, d) log_info("-- STEP %d: %s", n, d)
log_decision <- function(v, val, why) log_info("DECISION | %s = %s | %s", v, val, why)
dir.create("logs", recursive=TRUE, showWarnings=FALSE)

#
# Arguments:
#   species_name_or_list_csv : Species name (e.g., "Chelonia mydas") or path to CSV
#                              with column "scientificName"
#   output_dir               : Directory to write outputs (created if absent)
#   year_from                : Minimum year of observation (optional, default: 1950)
#   year_to                  : Maximum year of observation (optional, default: current year)
#   wkt_geometry             : WKT polygon to restrict query (optional, e.g., "POLYGON((-80 -30,-80 10,-30 10,-30 -30,-80 -30))")
#
# Outputs (per species):
#   occurrences_raw_OBIS_{species}_{date}.csv  — standardised occurrence records
#   download_metadata_OBIS_{species}.txt        — download provenance and citation
#
# Standard output schema:
#   species, decimalLatitude, decimalLongitude, eventDate, countryCode,
#   basisOfRecord, coordinateUncertaintyInMeters, datasetName, occurrenceID,
#   source, download_doi
# Extra OBIS columns:
#   depth, marine

suppressPackageStartupMessages(library(robis))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(readr))

# ── 1. Parse arguments ───────────────────────────────────────────────────────
log_step(1, "Parse command-line arguments")
args <- commandArgs(trailingOnly = TRUE)

if (length(args) < 2) {
  species_input <- "Chelonia mydas"
  output_dir    <- "output/obis"
  year_from     <- 1950
  year_to       <- as.integer(format(Sys.Date(), "%Y"))
  wkt_geometry  <- NULL
  log_warn("Fewer than 2 arguments provided. Using default values for testing.")
} else {
  species_input <- args[1]
  output_dir    <- args[2]
  year_from     <- if (length(args) >= 3) as.integer(args[3]) else 1950
  year_to       <- if (length(args) >= 4) as.integer(args[4]) else as.integer(format(Sys.Date(), "%Y"))
  wkt_geometry  <- if (length(args) >= 5 && args[5] != "") args[5] else NULL
}

log_info("Script: download_from_obis.R | Skill: %s", SKILL_NAME)
log_info("Species input  : %s", species_input)
log_info("Output dir     : %s", output_dir)
log_info("Year range     : %d - %d", year_from, year_to)
log_info("WKT geometry   : %s", ifelse(is.null(wkt_geometry), "none (global)", wkt_geometry))

log_decision("absence", "FALSE",
             "apenas registros de presenca confirmada; OBIS inclui dados de ausencia em alguns datasets")
log_decision("year_from", year_from, "temporal filter; 1950 covers the modern era of marine records")

# ── 2. Create output directory ───────────────────────────────────────────────
log_step(2, "Create output directory")
dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

# ── 3. Build species list ────────────────────────────────────────────────────
log_step(3, "Build species list")
if (grepl("\\.csv$", species_input, ignore.case = TRUE) && file.exists(species_input)) {
  tryCatch({
    species_df <- read_csv(species_input, show_col_types = FALSE)
    if (!"scientificName" %in% names(species_df)) {
      log_error(
        "Coluna 'scientificName' nao encontrada em: %s\nProbable cause: CSV mal formatado.\nPrevious skill: ecological-data-foundation",
        species_input
      )
      stop("Missing column 'scientificName'")
    }
    species_list <- unique(trimws(species_df$scientificName))
    log_info("Modo batch: %d especies carregadas", length(species_list))
    log_decision("mode", "batch", "valid CSV with scientificName column")
  }, error = function(e) {
    log_error(
      "Failed to read lista de especies: %s\nPrevious skill: ecological-data-foundation",
      conditionMessage(e)
    )
    stop(e)
  })
} else {
  species_list <- trimws(species_input)
  log_info("Modo especie unica: %s", species_list)
  log_decision("mode", "single_species", "argument is not a CSV file")
}

# ── 4. Download function ─────────────────────────────────────────────────────
download_obis_species <- function(sp_name) {
  log_info("--- Iniciando download OBIS: %s ---", sp_name)
  today_str <- format(Sys.Date(), "%Y%m%d")
  safe_name <- gsub(" ", "_", sp_name)

  # Build query
  query_args <- list(
    scientificname = sp_name,
    absence        = FALSE,
    startdate      = paste0(year_from, "-01-01"),
    enddate        = paste0(year_to,   "-12-31")
  )
  if (!is.null(wkt_geometry)) {
    query_args$geometry <- wkt_geometry
  }

  occ_raw <- tryCatch({
    do.call(robis::occurrence, query_args)
  }, error = function(e) {
    log_error(
      "Failed in robis::occurrence para '%s': %s\nProbable cause: no internet connection or OBIS API unavailable.\nCheck: https://api.obis.org/\nPrevious skill: ecological-data-foundation",
      sp_name, conditionMessage(e)
    )
    stop(e)
  })

  if (is.null(occ_raw) || nrow(occ_raw) == 0) {
    log_warn("No OBIS records found for '%s'.", sp_name)
    return(invisible(NULL))
  }

  n_raw <- nrow(occ_raw)
  log_info("Raw records recuperados: %d", n_raw)

  # ── Apply OBIS quality flags ───────────────────────────────────────────────
  # Remove records flagged as having coordinate issues
  qc_cols <- c("flags")
  if ("flags" %in% names(occ_raw)) {
    bad_flags <- c("NO_COORD", "ZERO_COORD", "ON_LAND", "DEPTH_EXCEEDS_BATH")
    occ_raw <- occ_raw[!grepl(paste(bad_flags, collapse = "|"),
                               occ_raw$flags, ignore.case = TRUE), ]
    log_info("Records after OBIS quality flags filter: %d", nrow(occ_raw))
  }

  # ── Standardise to output schema ──────────────────────────────────────────
  get_col <- function(df, ...) {
    cols <- c(...)
    for (col in cols) {
      if (col %in% names(df)) return(df[[col]])
    }
    return(rep(NA, nrow(df)))
  }

  std <- data.frame(
    species                       = sp_name,
    decimalLatitude               = as.numeric(get_col(occ_raw, "decimalLatitude", "latitude")),
    decimalLongitude              = as.numeric(get_col(occ_raw, "decimalLongitude", "longitude")),
    eventDate                     = as.character(get_col(occ_raw, "eventDate", "date_start")),
    countryCode                   = as.character(get_col(occ_raw, "countryCode", "country")),
    basisOfRecord                 = as.character(get_col(occ_raw, "basisOfRecord")),
    coordinateUncertaintyInMeters = as.numeric(get_col(occ_raw, "coordinateUncertaintyInMeters")),
    datasetName                   = as.character(get_col(occ_raw, "datasetName", "dataset_name")),
    occurrenceID                  = as.character(get_col(occ_raw, "occurrenceID", "id")),
    source                        = "OBIS",
    download_doi                  = NA_character_,
    depth                         = as.numeric(get_col(occ_raw, "depth")),
    marine                        = TRUE,
    stringsAsFactors              = FALSE
  )

  # Replace empty basisOfRecord
  std$basisOfRecord[is.na(std$basisOfRecord) | std$basisOfRecord == ""] <- "OCCURRENCE"

  # Remove records with missing coordinates
  n_before <- nrow(std)
  std <- std[!is.na(std$decimalLatitude) & !is.na(std$decimalLongitude), ]
  n_removed <- n_before - nrow(std)
  if (n_removed > 0) {
    log_warn("%d registros removidos por coordenadas missing.", n_removed)
  }

  n_final <- nrow(std)
  log_info("Records with valid coordinates: %d", n_final)

  if (n_final < 30) {
    log_warn(
      "Insufficient records for reliable analysis (n = %d). Consider relaxing date or area filters.",
      n_final
    )
  }

  # ── Save CSV ───────────────────────────────────────────────────────────────
  csv_path <- file.path(output_dir, paste0("occurrences_raw_OBIS_", safe_name, "_", today_str, ".csv"))
  tryCatch({
    write_csv(std, csv_path)
    log_info("Written: %s (%d registros)", csv_path, n_final)
  }, error = function(e) {
    log_error(
      "Failed to gravar CSV para '%s': %s\nPrevious skill: ecological-data-foundation",
      sp_name, conditionMessage(e)
    )
    stop(e)
  })

  # ── Save metadata ──────────────────────────────────────────────────────────
  meta_lines <- c(
    paste("Species:", sp_name),
    paste("Source: Ocean Biodiversity Information System (OBIS) — https://obis.org"),
    paste("API endpoint: https://api.obis.org/v3/occurrence"),
    paste("Absence records excluded: TRUE"),
    paste("OBIS quality flags applied: TRUE (NO_COORD, ZERO_COORD, ON_LAND, DEPTH_EXCEEDS_BATH removed)"),
    paste("Year range:", year_from, "-", year_to),
    paste("WKT geometry:", ifelse(is.null(wkt_geometry), "none (global)", wkt_geometry)),
    paste("n_records:", n_final),
    paste("Download date:", Sys.Date()),
    paste("Citation: OBIS (", format(Sys.Date(), "%Y"), ") Ocean Biodiversity Information System. Intergovernmental Oceanographic Commission of UNESCO. www.obis.org. Accessed on", Sys.Date()),
    paste("License: CC0 1.0 (https://creativecommons.org/publicdomain/zero/1.0/)")
  )
  meta_path <- file.path(output_dir, paste0("download_metadata_OBIS_", safe_name, ".txt"))
  tryCatch({
    writeLines(meta_lines, meta_path)
    log_info("Written: %s", meta_path)
  }, error = function(e) {
    log_error(
      "Failed to save metadata for '%s': %s\nPrevious skill: ecological-data-foundation",
      sp_name, conditionMessage(e)
    )
  })

  return(invisible(csv_path))
}

# ── 5. Run for all species ───────────────────────────────────────────────────
log_step(4, "Run OBIS download for all species")
for (sp in species_list) {
  tryCatch(
    download_obis_species(sp),
    error = function(e) {
      log_error(
        "Failed to download '%s' from OBIS: %s\nProbable cause: network error or species not found.\nPrevious skill: ecological-data-foundation",
        sp, conditionMessage(e)
      )
    }
  )
}

log_info("All OBIS downloads completed. Check: %s", output_dir)
